File: src/embeddings/openai.ts
import { mapPool, resolveWorkerConcurrency } from "../concurrency.ts";
import { envFirstString, envOptionalInt, resolveEmbeddingApiKey } from "../env.ts";

const DEFAULT_MODEL = "ds1-potion-code-16m";
const DEFAULT_BASE_URL = "https://infer.dev.takara.ai/v1";
const DEFAULT_BATCH_SIZE = 32;
const DEFAULT_MAX_EMBED_CHARS = 1300;
const WINDOW_OVERLAP_CHARS = 120;

/** Native output size for Takara ds1-potion-code-16m (override with MIRU_EMBEDDING_DIMENSIONS). */
const MODEL_DEFAULT_DIMENSIONS: Record<string, number> = {
  "ds1-potion-code-16m": 256,
};

export interface EmbeddingTransportStats {
  requests: number;
  retries: number;
  payloadTooLarge: number;
  errors: number;
  inputItems: number;
  inputChars: number;
  totalRttMs: number;
  maxRttMs: number;
}

export interface EmbeddingBackend {
  readonly model: string;
  readonly dimensions: number;
  embedDocuments(texts: string[]): Promise<Float32Array[]>;
  embedQuery(text: string): Promise<Float32Array>;
}

export function resolveEmbeddingModel(): string {
  return envFirstString(
    ["MIRU_OPENAI_EMBEDDING_MODEL", "SEMBLE_OPENAI_EMBEDDING_MODEL", "OPENAI_EMBEDDING_MODEL"],
    DEFAULT_MODEL,
  );
}

export function resolveMaxEmbedChars(): number {
  return (
    envOptionalInt(["MIRU_MAX_EMBED_CHARS", "SEMBLE_MAX_EMBED_CHARS"], 256) ??
    DEFAULT_MAX_EMBED_CHARS
  );
}

function splitIntoWindows(text: string, maxChars: number): string[] {
  if (text.length <= maxChars) {
    return [text];
  }
  const out: string[] = [];
  const step = Math.max(64, maxChars - WINDOW_OVERLAP_CHARS);
  for (let start = 0; start < text.length; start += step) {
    const end = Math.min(text.length, start + maxChars);
    const part = text.slice(start, end);
    if (part.length > 0) {
      out.push(part);
    }
    if (end >= text.length) {
      break;
    }
  }
  return out;
}

function sanitizeEmbeddingInput(text: string): string {
  // Some OpenAI-compatible gateways mis-handle backslash escapes in JSON text
  // payloads. Keep a benchmark fallback mode to forcefully neutralize them.
  const mode = process.env.MIRU_EMBED_ESCAPE_MODE ?? process.env.SEMBLE_EMBED_ESCAPE_MODE ?? "quad";
  if (mode === "strip") {
    return text.replaceAll("\\", "/");
  }
  return text.replaceAll("\\", "\\\\\\\\");
}

function isPayloadTooLargeError(err: unknown): boolean {
  const status = (err as { status?: number }).status;
  if (status === 413) {
    return true;
  }
  const message = (err as { message?: string }).message;
  if (typeof message === "string" && message.includes("413")) {
    return true;
  }
  return false;
}

export function resolveEmbeddingDimensions(model?: string): number | undefined {
  const fromEnv = envOptionalInt([
    "MIRU_EMBEDDING_DIMENSIONS",
    "SEMBLE_EMBEDDING_DIMENSIONS",
    "OPENAI_EMBEDDING_DIMENSIONS",
  ]);
  if (fromEnv != null) {
    return fromEnv;
  }
  return MODEL_DEFAULT_DIMENSIONS[model ?? resolveEmbeddingModel()];
}

export function resolveEmbeddingBatchSize(): number {
  return (
    envOptionalInt([
      "MIRU_EMBEDDING_BATCH_SIZE",
      "SEMBLE_EMBEDDING_BATCH_SIZE",
      "OPENAI_EMBEDDING_BATCH_SIZE",
    ]) ?? DEFAULT_BATCH_SIZE
  );
}

export function resolveEmbeddingBaseUrl(): string {
  return envFirstString(
    ["MIRU_OPENAI_BASE_URL", "SEMBLE_OPENAI_BASE_URL", "OPENAI_BASE_URL"],
    DEFAULT_BASE_URL,
  ).replace(/\/$/, "");
}

interface EmbeddingResponseItem {
  index: number;
  embedding: number[];
}

interface EmbeddingResponse {
  data: EmbeddingResponseItem[];
}

interface EmbeddingClient {
  createEmbeddings(
    input: string[] | string,
    model: string,
    dimensions?: number,
  ): Promise<EmbeddingResponse>;
}

class EmbeddingApiError extends Error {
  readonly status: number;
  readonly body: string;

  constructor(status: number, body: string) {
    super(`Embedding API error ${status}: ${body.slice(0, 500)}`);
    this.name = "EmbeddingApiError";
    this.status = status;
    this.body = body;
  }
}

function createClient(): EmbeddingClient {
  const apiKey = resolveEmbeddingApiKey();

  const baseUrl = resolveEmbeddingBaseUrl();
  const endpoint = `${baseUrl}/embeddings`;

  return {
    async createEmbeddings(
      input: string[] | string,
      model: string,
      dimensions?: number,
    ): Promise<EmbeddingResponse> {
      const body: Record<string, unknown> = { model, input };
      if (dimensions != null) {
        body.dimensions = dimensions;
      }
      const response = await fetch(endpoint, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        throw new EmbeddingApiError(response.status, await response.text());
      }

      const payload = (await response.json()) as EmbeddingResponse;
      if (!payload || !Array.isArray(payload.data)) {
        throw new Error("Embedding API returned invalid payload");
      }
      return payload;
    },
  };
}

export class OpenAIEmbeddingBackend implements EmbeddingBackend {
  readonly model: string;
  dimensions = 0;

  private readonly client: EmbeddingClient;
  private readonly batchSize: number;
  private readonly maxEmbedChars: number;
  private readonly requestedDimensions: number | undefined;
  private stats: EmbeddingTransportStats;

  constructor(options?: {
    model?: string;
    batchSize?: number;
    maxEmbedChars?: number;
    dimensions?: number;
    client?: EmbeddingClient;
  }) {
    this.model = options?.model ?? resolveEmbeddingModel();
    this.client = options?.client ?? createClient();
    this.batchSize = options?.batchSize ?? resolveEmbeddingBatchSize();
    this.maxEmbedChars = options?.maxEmbedChars ?? resolveMaxEmbedChars();
    this.requestedDimensions = options?.dimensions ?? resolveEmbeddingDimensions(this.model);
    this.stats = {
      requests: 0,
      retries: 0,
      payloadTooLarge: 0,
      errors: 0,
      inputItems: 0,
      inputChars: 0,
      totalRttMs: 0,
      maxRttMs: 0,
    };
  }

  async embedDocuments(texts: string[]): Promise<Float32Array[]> {
    if (texts.length === 0) {
      return [];
    }

    interface WindowJob {
      docIndex: number;
      text: string;
    }

    const jobs: WindowJob[] = [];
    const windowVectors: Float32Array[][] = texts.map(() => []);

    for (let docIndex = 0; docIndex < texts.length; docIndex++) {
      const text = texts[docIndex];
      if (text === undefined) {
        continue;
      }
      const windows = splitIntoWindows(text, this.maxEmbedChars).map(sanitizeEmbeddingInput);
      for (const text of windows) {
        jobs.push({ docIndex, text });
      }
    }

    if (jobs.length === 0) {
      return texts.map(() => new Float32Array(0));
    }

    const batches: WindowJob[][] = [];
    for (let i = 0; i < jobs.length; i += this.batchSize) {
      batches.push(jobs.slice(i, i + this.batchSize));
    }

    const concurrency = resolveWorkerConcurrency();
    await mapPool(batches, concurrency, async (batch) => {
      const vectors = await this.embedBatchRawWithRetry(batch.map((job) => job.text));
      for (let i = 0; i < batch.length; i++) {
        const job = batch[i];
        const vector = vectors[i];
        if (job === undefined || vector === undefined) {
          continue;
        }
        const bucket = windowVectors[job.docIndex];
        if (bucket) {
          bucket.push(vector);
        }
      }
    });

    const out = windowVectors.map((vectors) => poolWindowVectors(vectors));
    for (const vec of out) {
      if (this.dimensions === 0 && vec.length > 0) {
        this.dimensions = vec.length;
      }
    }
    return out;
  }

  private async embedBatchRawWithRetry(texts: string[]): Promise<Float32Array[]> {
    if (texts.length === 0) {
      return [];
    }
    try {
      const started = performance.now();
      const response = await this.client.createEmbeddings(
        texts,
        this.model,
        this.requestedDimensions,
      );
      const elapsed = performance.now() - started;
      this.stats.requests += 1;
      this.stats.inputItems += texts.length;
      this.stats.inputChars += texts.reduce((acc, text) => acc + text.length, 0);
      this.stats.totalRttMs += elapsed;
      this.stats.maxRttMs = Math.max(this.stats.maxRttMs, elapsed);
      const vectors = vectorsFromResponse(response.data, texts.length);
      for (const vec of vectors) {
        if (this.requestedDimensions != null && vec.length !== this.requestedDimensions) {
          throw new Error(
            `Embedding API returned ${vec.length} dims for model ${this.model}, expected ${this.requestedDimensions}`,
          );
        }
        if (this.dimensions === 0) {
          this.dimensions = vec.length;
        } else if (vec.length !== this.dimensions) {
          throw new Error(
            `Inconsistent embedding dimensions in batch: ${vec.length} vs ${this.dimensions}`,
          );
        }
      }
      return vectors;
    } catch (err: unknown) {
      if (isPayloadTooLargeError(err) && texts.length > 1) {
        this.stats.payloadTooLarge += 1;
        this.stats.retries += 1;
        const mid = Math.ceil(texts.length / 2);
        const left = await this.embedBatchRawWithRetry(texts.slice(0, mid));
        const right = await this.embedBatchRawWithRetry(texts.slice(mid));
        return [...left, ...right];
      }
      if (isPayloadTooLargeError(err) && texts.length === 1) {
        this.stats.payloadTooLarge += 1;
        this.stats.retries += 1;
        const text = texts[0];
        if (!text) {
          this.stats.errors += 1;
          throw err;
        }
        if (text.length <= 128) {
          this.stats.errors += 1;
          throw err;
        }
        const mid = Math.floor(text.length / 2);
        const leftText = text.slice(0, mid);
        const rightText = text.slice(mid);
        const left = await this.embedBatchRawWithRetry([leftText]);
        const right = await this.embedBatchRawWithRetry([rightText]);
        return [...left, ...right];
      }
      this.stats.errors += 1;
      throw err;
    }
  }

  async embedQuery(text: string): Promise<Float32Array> {
    const [vec] = await this.embedDocuments([text]);
    if (!vec) {
      throw new Error("OpenAI returned no embedding for query");
    }
    return vec;
  }

  getStats(): EmbeddingTransportStats {
    return { ...this.stats };
  }

  resetStats(): void {
    this.stats = {
      requests: 0,
      retries: 0,
      payloadTooLarge: 0,
      errors: 0,
      inputItems: 0,
      inputChars: 0,
      totalRttMs: 0,
      maxRttMs: 0,
    };
  }
}

function vectorsFromResponse(data: EmbeddingResponseItem[], expected: number): Float32Array[] {
  const byIndex = new Map<number, Float32Array>();
  for (const item of data) {
    if (item.index >= 0 && item.index < expected) {
      byIndex.set(item.index, new Float32Array(item.embedding));
    }
  }
  if (byIndex.size !== expected) {
    throw new Error(
      `Embedding API returned ${data.length} vectors for ${expected} inputs (${byIndex.size} unique indices)`,
    );
  }
  return Array.from({ length: expected }, (_, i) => {
    const vec = byIndex.get(i);
    if (!vec) {
      throw new Error(`Missing embedding vector at index ${i}`);
    }
    return vec;
  });
}

function poolWindowVectors(vectors: Float32Array[]): Float32Array {
  if (vectors.length === 0) {
    throw new Error("Embedding API returned no vectors");
  }
  const first = vectors[0];
  if (!first) {
    throw new Error("Embedding API returned no vectors");
  }
  if (vectors.length === 1) {
    return normalize(first);
  }
  const pooled = new Float32Array(first.length);
  for (const vec of vectors) {
    for (let i = 0; i < pooled.length; i++) {
      pooled[i] += vec[i] ?? 0;
    }
  }
  for (let i = 0; i < pooled.length; i++) {
    pooled[i] /= vectors.length;
  }
  return normalize(pooled);
}

function normalize(vec: Float32Array): Float32Array {
  let norm = 0;
  for (let i = 0; i < vec.length; i++) {
    const value = vec[i] ?? 0;
    norm += value * value;
  }
  norm = Math.sqrt(norm);
  if (norm === 0) {
    return vec;
  }
  const out = new Float32Array(vec.length);
  for (let i = 0; i < vec.length; i++) {
    out[i] = (vec[i] ?? 0) / norm;
  }
  return out;
}

let defaultBackend: OpenAIEmbeddingBackend | null = null;

export function getEmbeddingBackend(model?: string): OpenAIEmbeddingBackend {
  if (model) {
    return new OpenAIEmbeddingBackend({
      model,
      dimensions: resolveEmbeddingDimensions(model),
    });
  }
  if (!defaultBackend) {
    defaultBackend = new OpenAIEmbeddingBackend();
  }
  return defaultBackend;
}

File: internal/embeddings/openai.go
package embeddings

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"os"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/takara-ai/miru-code/internal/env"
)

const (
	defaultModel         = "ds1-potion-code-16m"
	defaultBaseURL       = "https://infer.dev.takara.ai/v1"
	defaultBatchSize     = 32
	defaultMaxEmbedChars = 1300
	windowOverlapChars   = 120
)

var modelDefaultDimensions = map[string]int{
	"ds1-potion-code-16m": 256,
}

type Backend interface {
	ModelName() string
	Dimensions() int
	EmbedDocuments(texts []string) ([][]float32, error)
	EmbedQuery(text string) ([]float32, error)
}

type TransportStats struct {
	Requests        int
	Retries         int
	PayloadTooLarge int
	Errors          int
	InputItems      int
	InputChars      int
	TotalRTTMS      float64
	MaxRTTMS        float64
}

func ResolveEmbeddingModel() string {
	return env.FirstString([]string{"MIRU_OPENAI_EMBEDDING_MODEL", "SEMBLE_OPENAI_EMBEDDING_MODEL", "OPENAI_EMBEDDING_MODEL"}, defaultModel)
}

func ResolveMaxEmbedChars() int {
	if value, ok := env.OptionalInt([]string{"MIRU_MAX_EMBED_CHARS", "SEMBLE_MAX_EMBED_CHARS"}, 256); ok {
		return value
	}
	return defaultMaxEmbedChars
}

func ResolveEmbeddingDimensions(model ...string) (int, bool) {
	if value, ok := env.OptionalInt([]string{"MIRU_EMBEDDING_DIMENSIONS", "SEMBLE_EMBEDDING_DIMENSIONS", "OPENAI_EMBEDDING_DIMENSIONS"}); ok {
		return value, true
	}
	resolved := ResolveEmbeddingModel()
	if len(model) > 0 && model[0] != "" {
		resolved = model[0]
	}
	value, ok := modelDefaultDimensions[resolved]
	return value, ok
}

func ResolveEmbeddingBatchSize() int {
	if value, ok := env.OptionalInt([]string{"MIRU_EMBEDDING_BATCH_SIZE", "SEMBLE_EMBEDDING_BATCH_SIZE", "OPENAI_EMBEDDING_BATCH_SIZE"}); ok {
		return value
	}
	return defaultBatchSize
}

func ResolveEmbeddingBaseURL() string {
	return strings.TrimRight(env.FirstString([]string{"MIRU_OPENAI_BASE_URL", "SEMBLE_OPENAI_BASE_URL", "OPENAI_BASE_URL"}, defaultBaseURL), "/")
}

func SanitizeEmbeddingInput(text string) string {
	out := strings.ToValidUTF8(text, "\uFFFD")
	mode := os.Getenv("MIRU_EMBED_ESCAPE_MODE")
	if mode == "" {
		mode = os.Getenv("SEMBLE_EMBED_ESCAPE_MODE")
	}
	if mode == "" {
		mode = "quad"
	}
	if mode == "strip" {
		return strings.ReplaceAll(out, "\\", "/")
	}
	return strings.ReplaceAll(out, "\\", "\\\\\\\\")
}

type Client interface {
	CreateEmbeddings(input []string, model string, dimensions *int) (EmbeddingResponse, error)
}

type EmbeddingResponseItem struct {
	Index     int       `json:"index"`
	Embedding []float64 `json:"embedding"`
}

type EmbeddingResponse struct {
	Data []EmbeddingResponseItem `json:"data"`
}

type APIError struct {
	Status int
	Body   string
}

func (e APIError) Error() string {
	body := e.Body
	if len(body) > 500 {
		body = body[:500]
	}
	return fmt.Sprintf("Embedding API error %d: %s", e.Status, body)
}

type HTTPClient struct {
	APIKey     string
	BaseURL    string
	HTTPClient *http.Client
}

func NewHTTPClient() (*HTTPClient, error) {
	key, err := env.ResolveEmbeddingAPIKey()
	if err != nil {
		return nil, err
	}
	return &HTTPClient{
		APIKey:     key,
		BaseURL:    ResolveEmbeddingBaseURL(),
		HTTPClient: http.DefaultClient,
	}, nil
}

func (c *HTTPClient) CreateEmbeddings(input []string, model string, dimensions *int) (EmbeddingResponse, error) {
	body := map[string]any{
		"model": model,
		"input": input,
	}
	if dimensions != nil {
		body["dimensions"] = *dimensions
	}
	data, err := json.Marshal(body)
	if err != nil {
		return EmbeddingResponse{}, err
	}

	req, err := http.NewRequest(http.MethodPost, strings.TrimRight(c.BaseURL, "/")+"/embeddings", bytes.NewReader(data))
	if err != nil {
		return EmbeddingResponse{}, err
	}
	req.Header.Set("Authorization", "Bearer "+c.APIKey)
	req.Header.Set("Content-Type", "application/json")

	client := c.HTTPClient
	if client == nil {
		client = http.DefaultClient
	}
	resp, err := client.Do(req)
	if err != nil {
		return EmbeddingResponse{}, err
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return EmbeddingResponse{}, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return EmbeddingResponse{}, APIError{Status: resp.StatusCode, Body: string(raw)}
	}

	var payload EmbeddingResponse
	if err := json.Unmarshal(raw, &payload); err != nil {
		return EmbeddingResponse{}, err
	}
	if payload.Data == nil {
		return EmbeddingResponse{}, fmt.Errorf("Embedding API returned invalid payload")
	}
	return payload, nil
}

type OpenAIBackend struct {
	model               string
	dimensions          int
	client              Client
	batchSize           int
	maxEmbedChars       int
	requestedDimensions *int
	stats               TransportStats
}

type Options struct {
	Model      string
	BatchSize  int
	MaxChars   int
	Dimensions *int
	Client     Client
}

func NewOpenAIBackend(options Options) (*OpenAIBackend, error) {
	model := options.Model
	if model == "" {
		model = ResolveEmbeddingModel()
	}
	client := options.Client
	var err error
	if client == nil {
		client, err = NewHTTPClient()
		if err != nil {
			return nil, err
		}
	}
	batchSize := options.BatchSize
	if batchSize == 0 {
		batchSize = ResolveEmbeddingBatchSize()
	}
	maxChars := options.MaxChars
	if maxChars == 0 {
		maxChars = ResolveMaxEmbedChars()
	}
	dimensions := options.Dimensions
	if dimensions == nil {
		if value, ok := ResolveEmbeddingDimensions(model); ok {
			copy := value
			dimensions = &copy
		}
	}
	return &OpenAIBackend{
		model:               model,
		client:              client,
		batchSize:           batchSize,
		maxEmbedChars:       maxChars,
		requestedDimensions: dimensions,
	}, nil
}

func (b *OpenAIBackend) ModelName() string {
	return b.model
}

func (b *OpenAIBackend) Dimensions() int {
	return b.dimensions
}

func (b *OpenAIBackend) EmbedDocuments(texts []string) ([][]float32, error) {
	if len(texts) == 0 {
		return nil, nil
	}

	type windowJob struct {
		docIndex int
		text     string
	}

	jobs := []windowJob{}
	windowVectors := make([][][]float32, len(texts))
	for docIndex, text := range texts {
		for _, window := range splitIntoWindows(text, b.maxEmbedChars) {
			jobs = append(jobs, windowJob{docIndex: docIndex, text: SanitizeEmbeddingInput(window)})
		}
	}
	if len(jobs) == 0 {
		return make([][]float32, len(texts)), nil
	}

	for i := 0; i < len(jobs); i += b.batchSize {
		end := i + b.batchSize
		if end > len(jobs) {
			end = len(jobs)
		}
		batch := jobs[i:end]
		input := make([]string, len(batch))
		for j, job := range batch {
			input[j] = job.text
		}
		vectors, err := b.embedBatchRawWithRetry(input)
		if err != nil {
			return nil, err
		}
		for j, job := range batch {
			if j < len(vectors) {
				windowVectors[job.docIndex] = append(windowVectors[job.docIndex], vectors[j])
			}
		}
	}

	out := make([][]float32, len(texts))
	for i, vectors := range windowVectors {
		pooled, err := poolWindowVectors(vectors)
		if err != nil {
			return nil, err
		}
		if b.dimensions == 0 && len(pooled) > 0 {
			b.dimensions = len(pooled)
		}
		out[i] = pooled
	}
	return out, nil
}

func (b *OpenAIBackend) EmbedQuery(text string) ([]float32, error) {
	vectors, err := b.EmbedDocuments([]string{text})
	if err != nil {
		return nil, err
	}
	if len(vectors) == 0 {
		return nil, fmt.Errorf("OpenAI returned no embedding for query")
	}
	return vectors[0], nil
}

func (b *OpenAIBackend) Stats() TransportStats {
	return b.stats
}

func (b *OpenAIBackend) ResetStats() {
	b.stats = TransportStats{}
}

func (b *OpenAIBackend) embedBatchRawWithRetry(texts []string) ([][]float32, error) {
	if len(texts) == 0 {
		return nil, nil
	}
	started := time.Now()
	response, err := b.client.CreateEmbeddings(texts, b.model, b.requestedDimensions)
	if err == nil {
		elapsed := float64(time.Since(started).Microseconds()) / 1000
		b.stats.Requests++
		b.stats.InputItems += len(texts)
		for _, text := range texts {
			b.stats.InputChars += utf8.RuneCountInString(text)
		}
		b.stats.TotalRTTMS += elapsed
		if elapsed > b.stats.MaxRTTMS {
			b.stats.MaxRTTMS = elapsed
		}
		vectors, err := vectorsFromResponse(response.Data, len(texts))
		if err != nil {
			b.stats.Errors++
			return nil, err
		}
		for _, vec := range vectors {
			if b.requestedDimensions != nil && len(vec) != *b.requestedDimensions {
				return nil, fmt.Errorf("Embedding API returned %d dims for model %s, expected %d", len(vec), b.model, *b.requestedDimensions)
			}
			if b.dimensions == 0 {
				b.dimensions = len(vec)
			} else if len(vec) != b.dimensions {
				return nil, fmt.Errorf("Inconsistent embedding dimensions in batch: %d vs %d", len(vec), b.dimensions)
			}
		}
		return vectors, nil
	}

	if isPayloadTooLargeError(err) && len(texts) > 1 {
		b.stats.PayloadTooLarge++
		b.stats.Retries++
		mid := (len(texts) + 1) / 2
		left, leftErr := b.embedBatchRawWithRetry(texts[:mid])
		if leftErr != nil {
			return nil, leftErr
		}
		right, rightErr := b.embedBatchRawWithRetry(texts[mid:])
		if rightErr != nil {
			return nil, rightErr
		}
		return append(left, right...), nil
	}
	if isPayloadTooLargeError(err) && len(texts) == 1 {
		b.stats.PayloadTooLarge++
		b.stats.Retries++
		text := texts[0]
		if len(text) <= 128 {
			b.stats.Errors++
			return nil, err
		}
		mid := len(text) / 2
		left, leftErr := b.embedBatchRawWithRetry([]string{text[:mid]})
		if leftErr != nil {
			return nil, leftErr
		}
		right, rightErr := b.embedBatchRawWithRetry([]string{text[mid:]})
		if rightErr != nil {
			return nil, rightErr
		}
		return append(left, right...), nil
	}
	b.stats.Errors++
	return nil, err
}

func splitIntoWindows(text string, maxChars int) []string {
	if len(text) <= maxChars {
		return []string{text}
	}
	out := []string{}
	step := max(64, maxChars-windowOverlapChars)
	for start := 0; start < len(text); start += step {
		end := min(len(text), start+maxChars)
		part := text[start:end]
		if len(part) > 0 {
			out = append(out, part)
		}
		if end >= len(text) {
			break
		}
	}
	return out
}

func isPayloadTooLargeError(err error) bool {
	if apiErr, ok := err.(APIError); ok && apiErr.Status == http.StatusRequestEntityTooLarge {
		return true
	}
	return strings.Contains(err.Error(), "413")
}

func vectorsFromResponse(data []EmbeddingResponseItem, expected int) ([][]float32, error) {
	byIndex := map[int][]float32{}
	for _, item := range data {
		if item.Index >= 0 && item.Index < expected {
			vec := make([]float32, len(item.Embedding))
			for i, value := range item.Embedding {
				vec[i] = float32(value)
			}
			byIndex[item.Index] = vec
		}
	}
	if len(byIndex) != expected {
		return nil, fmt.Errorf("Embedding API returned %d vectors for %d inputs (%d unique indices)", len(data), expected, len(byIndex))
	}
	out := make([][]float32, expected)
	for i := 0; i < expected; i++ {
		vec, ok := byIndex[i]
		if !ok {
			return nil, fmt.Errorf("Missing embedding vector at index %d", i)
		}
		out[i] = vec
	}
	return out, nil
}

func poolWindowVectors(vectors [][]float32) ([]float32, error) {
	if len(vectors) == 0 {
		return nil, fmt.Errorf("Embedding API returned no vectors")
	}
	if len(vectors) == 1 {
		return normalize(vectors[0]), nil
	}
	pooled := make([]float32, len(vectors[0]))
	for _, vec := range vectors {
		for i := range pooled {
			if i < len(vec) {
				pooled[i] += vec[i]
			}
		}
	}
	for i := range pooled {
		pooled[i] /= float32(len(vectors))
	}
	return normalize(pooled), nil
}

func normalize(vec []float32) []float32 {
	var norm float64
	for _, value := range vec {
		norm += float64(value) * float64(value)
	}
	norm = math.Sqrt(norm)
	if norm == 0 {
		return vec
	}
	out := make([]float32, len(vec))
	for i, value := range vec {
		out[i] = float32(float64(value) / norm)
	}
	return out
}
